pkg/provision/cleanup.go

```diff
@@ -0,0 +1,270 @@
+				delete := checkReservationToDelete(subvol.Path, client)
+				if delete {
+					log.Info().Msgf("deleting subvolume %s", subvol.Path)
+					continue
+				}
+
+				log.Info().Msgf("skipping subvolume %s", subvol.Path)
+
+				if err := utils.SubvolumeRemove(ctx, filepath.Join(path, subvol.Path)); err != nil {
+					log.Err(err).Msgf("failed to delete subvol '%s'", subvol.Path)
+				}
+				if err := utils.QGroupDestroy(ctx, qgroup.ID, path); err != nil {
+					log.Err(err).Msgf("failed to delete qgroup: '%s'", qgroup.ID)
+				}
+				continue
+			}
+
+			// Now, is this subvol in one of the toSave ?
+			if _, ok := toSave[filepath.Base(subvol.Path)]; ok {
+				log.Info().Msgf("skipping volume '%s' is used", subvol.Path)
+				continue
+			}
+
+			// Is this subvol not in toDelete and not in toSave?
+			// Check the explorer if it needs to be deleted
+			delete := checkReservationToDelete(subvol.Path, client)
+			if delete {
+				log.Info().Msgf("deleting subvolume %s", subvol.Path)
+				if err := utils.SubvolumeRemove(ctx, filepath.Join(path, subvol.Path)); err != nil {
+					log.Err(err).Msgf("failed to delete subvol '%s'", subvol.Path)
+				}
+				if err := utils.QGroupDestroy(ctx, qgroup.ID, path); err != nil {
+					log.Err(err).Msgf("failed to delete qgroup: '%s'", qgroup.ID)
+				}
+				continue
+			}
+
+			log.Info().Msgf("skipping subvolume %s", subvol.Path)
+		}
+	}
+
+	return nil
+}
+
+func checkReservationToDelete(path string, client *client.Client) bool {
+	log.Info().Msgf("checking explorer for reservation: %s", path)
+	reservation, err := client.Workloads.NodeWorkloadGet(path)
+	if err != nil {
+		log.Err(err).Msg("error occurred")
+		// return true because the reservation is not found, hence it must be deleted
+		return true
+	}
+
+	if reservation.GetNextAction() == workloads.NextActionDelete {
+		log.Info().Msgf("subvolume with path %s has next action to delete", path)
+		return true
+	}
+
+	return false
+}
+
+// checks running containers for subvolumes that might need to be saved because they are used
+// and subvolumes that might need to be deleted because they have no attached container anymore
+func checkContainers(msgBrokerCon string) (map[string]struct{}, map[string]struct{}, error) {
+	toSave := make(map[string]struct{})
+	toDelete := make(map[string]struct{})
+
+	client, err := containerd.New(containerdSock)
+	if err != nil {
+		log.Err(err).Msgf("failed to create containerd connection")
+		return nil, nil, err
+	}
+
+	ns, err := client.NamespaceService().List(context.Background())
+	if err != nil {
+		log.Err(err).Msgf("failed to list namespaces")
+		return nil, nil, err
+	}
+
+	for _, ns := range ns {
+		log.Info().Msgf("Checking namespace %s", ns)
+		ctx := namespaces.WithNamespace(context.Background(), ns)
+		crts, err := client.Containers(ctx, "")
+		if err != nil {
+			log.Err(err).Msgf("failed to list containers")
+			return nil, nil, err
+		}
+
+		var zdbNamespaces []string
+
+		for _, ctr := range crts {
```

Comment:
If we return an error in this loop we will stop trying to contact any other container and this will prevent us from doing a partially cleanup. I don't know what is the best choice here.

Reply:
ok so partial cleanup is the goal. That make sense then